sound/soc/sof/probe.c

```diff
@@ -0,0 +1,290 @@
+ * are no probes connected and with all injectors detached.
+ */
+int sof_ipc_probe_deinit(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc_cmd_hdr msg;
+	struct sof_ipc_reply reply;
+
+	msg.size = sizeof(msg);
+	msg.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DEINIT;
+
+	return sof_ipc_tx_message(sdev->ipc, msg.cmd, &msg, msg.size,
+			&reply, sizeof(reply));
+}
+EXPORT_SYMBOL(sof_ipc_probe_deinit);
+
+static int sof_ipc_probe_get(struct snd_sof_dev *sdev, unsigned int cmd,
+		void **params, size_t *bytes)
+{
+	struct sof_ipc_probe_get_params msg = {0};
+	struct sof_ipc_probe_get_params *reply;
+	size_t size;
+	int ret;
+
+	*params = NULL;
+	*bytes = 0;
+
+	reply = kzalloc(SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	if (!reply)
+		return -ENOMEM;
+	msg.rhdr.hdr.size = sizeof(msg);
+	msg.rhdr.hdr.cmd = SOF_IPC_GLB_PROBE | cmd;
+
+	ret = sof_ipc_tx_message(sdev->ipc, msg.rhdr.hdr.cmd, &msg,
+			msg.rhdr.hdr.size, reply, SOF_IPC_MSG_MAX_SIZE);
+	if (ret < 0 || reply->rhdr.error < 0)
+		goto exit;
+
+	size = reply->rhdr.hdr.size -= sizeof(*reply);
+	if (size <= 0)
+		goto exit;
+
+	*params = kmemdup(&reply->dma[0], size, GFP_KERNEL);
+	if (!*params) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+	*bytes = size;
+
+exit:
+	kfree(reply);
+	return ret;
+}
+
+/**
+ * sof_ipc_probe_get_dma - retrieve list of active injection dmas
+ * @sdev:	SOF sound device
+ * @dma:	Returned list of active dmas
+ * @num_dma:	Returned count of active dmas
+ *
+ * Host sends DMA_GET request to obtain list of injection dmas it
+ * can use to transfer data over with.
+ *
+ * Note that list contains only injection dmas as there is only one
+ * extractor (dma) and it is always assigned on probing init.
+ * DSP knows exactly where data from extraction probes is going to,
+ * which is not the case for injection where multiple streams
+ * could be engaged.
+ */
+int sof_ipc_probe_get_dma(struct snd_sof_dev *sdev,
+		struct sof_probe_dma **dma, size_t *num_dma)
+{
+	int ret;
+
+	ret = sof_ipc_probe_get(sdev, SOF_IPC_PROBE_DMA_GET,
+			(void **)dma, num_dma);
+	*num_dma /= sizeof(**dma);
+	return ret;
+}
+EXPORT_SYMBOL(sof_ipc_probe_get_dma);
+
+/**
+ * sof_ipc_probe_dma_attach - attach to specified dmas
+ * @sdev:	SOF sound device
+ * @dma:	List of streams (dmas) to attach to
+ * @num_dma:	Number of elements in @dma
+ *
+ * Contrary to extraction, injection streams are never assigned
+ * on init. Before attempting any data injection, host is responsible
+ * for specifying streams which will be later used to transfer data
+ * to connected probe points.
+ */
+int sof_ipc_probe_dma_attach(struct snd_sof_dev *sdev,
+		struct sof_probe_dma *dma, size_t num_dma)
+{
+	struct sof_ipc_probe_dma_set_params *msg;
+	struct sof_ipc_reply reply;
+	size_t bytes = sizeof(*dma) * num_dma;
+	int ret;
+
+	msg = kmalloc(sizeof(*msg) + bytes, GFP_KERNEL);
+	if (!msg)
+		return -ENOMEM;
+	msg->hdr.size = sizeof(*msg) + bytes;
+	msg->hdr.cmd = SOF_IPC_GLB_PROBE | SOF_IPC_PROBE_DMA_SET;
```

Comment:
is there a reason why a DMA_SET is wrapped as 'dma_attach'?

naming conventions do matter here.